tests/test_null_byte_insertion.py

```diff
@@ -0,0 +1,43 @@
+# Copyright 2021 The Matrix.org Foundation C.I.C.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import synapse.rest.admin
+from synapse.rest.client import account, login, register, room
+
+from tests.unittest import HomeserverTestCase
+
+
+class NullByteInsertionTest(HomeserverTestCase):
+    servlets = [
+        synapse.rest.admin.register_servlets_for_client_rest_resource,
+        login.register_servlets,
+        account.register_servlets,
+        register.register_servlets,
+        room.register_servlets,
+    ]
+
+    def setUp(self):
+        super().setUp()
+
+    # Note that this test must be run with postgres or else is meaningless,
+    # as sqlite will accept insertion of null code points
+    def test_null_byte(self):
+        self.register_user("alice", "password")
+        access_token = self.login("alice", "password")
+        room_id = self.helper.create_room_as("alice", True, "1", access_token)
+        body = '{"body":"\u0000", "msgtype":"m.text"}'
+
+        resp = self.helper.send(room_id, body, "1", access_token)
+        self.assertTrue("event_id" in resp)
```

Comment:
We should probably test searching for this content afterward to ensure that everything is working as we expect. (That probably means we'll need to expand the body to something that has words we can search for!)

Reply:
So I tried to test search using the `search_msgs` function https://github.com/matrix-org/synapse/blob/3eba047d388fd0d798229a0779f343dbda8a2887/synapse/storage/databases/main/search.py#L346 to search for text that was inserted (I changed the body to 'hi\u0000bob' to test that) but ran into problems because `search_msg` calls `_find_highlights_in_postgres` https://github.com/matrix-org/synapse/blob/3eba047d388fd0d798229a0779f343dbda8a2887/synapse/storage/databases/main/search.py#L623 which afaict pulls message bodies from different tables than the one this PR alters (`event_search`), and as such those message bodies still contain the nefarious `\u0000`, which causes `_find_highlights_in_postgres` to fail at this line: https://github.com/matrix-org/synapse/blob/3eba047d388fd0d798229a0779f343dbda8a2887/synapse/storage/databases/main/search.py#L675 causing the whole search to fail. This, however, seems out of the scope of this PR, since the idea was to specifically only alter the `event_search` table.
I did write a test that manually queries the `event_search` table to be sure that the string was still being stored and could be retrieved, would this be sufficient or should I take another approach?

Reply:
I think we ought to do something about `_find_highlights_in_postgres` — does this mean that if someone inserted a message containing a null byte, that they'd be able to prevent other users from performing any room searches that match the null-containing message? That sounds like a nasty bug, sort of a security issue.

I think replacing null bytes with spaces in `values` there also would make sense.

It might stop matching parts of that specific message (as in, it might not yield a result if the message says `"the\u0000mat"` and you search for `"the mat"`), but at least it won't break the search functionality entirely whenever a potential result contains a NULL and causes an exception that bubbles all the way back up to cause a 500 internal server error (if that makes sense?).

Reply:
(or in other words,

> causing the whole search to fail

is exactly the kind of thing we're interested in testing *doesn't* happen! :))

Reply:
> — does this mean that if someone inserted a message containing a null byte, that they'd be able to prevent other users from performing any room searches that match the null-containing message?

It does exactly mean that. It could also have been happening this whole time: if someone sent a "hi\u0000bob" and someone else sent "hi bob" then theoretically the first message body could be pulled up and cause things to fall down even though technically it was never stored in the `event_search` table because now the same "hi" and "bob" keywords have been stored. So I have been trying to work out whether that behavior predates/will persist after we fix search. Afaict just because message bodies were getting bounced out of the `events_search` table doesn't necessarily mean they were getting bounced out of cache, etc (as the testing shows)-that really depends on the order in which things are stored as the message comes in, doesn't it?

Reply:
> if someone sent a "hi\u0000bob" and someone else sent "hi bob" then theoretically the first message body could be pulled up

This doesn't match my understanding, because as you say, `"hi\u0000bob"` won't make it to the `event_search` table (due to the exact bug you're fixing in this PR).
If it doesn't make it into that table, then a search is not able to retrieve it.
If a search doesn't retrieve it, then that search won't ask Postgres to highlight it.

If you see some other way that this could happen, you may well be right — I'm not all that familiar! (so do let me know if this doesn't match up with what you expect)

(BTW, I don't think the 'highlighting' thing is obvious unless you have a clear picture of what the message search looks like — try searching a room if you haven't seen it yet. The only point of it seems to be to figure out *which* part of the message caused Postgres to match it.)